R/drop_dir.R

```diff
@@ -1,102 +1,126 @@
-                  include_membership = include_membership)))
-  metadata_url <- "https://api.dropbox.com/1/metadata/auto/"
-  if(!is.null(path)) {
-        metadata_url <- paste0(metadata_url, path)
-  }
-  req <- httr::GET(metadata_url, query = args, config(token = dtoken))
-  res <- jsonlite::fromJSON(httr::content(req, as = "text"), flatten = TRUE)
-  if(length(res$contents) > 0) { # i.e. not an empty folder
-  if(verbose) {
-    res2 <- res
-    res2$contents <- data.frame()
-    res$contents <- dplyr::tbl_df(res$contents)
-    pretty_lists(res2)
-    print(dplyr::tbl_df(res$contents))
-    invisible(res)
-  } else {
-    path <- mime_type <- root <- bytes <- modified <- NULL
-    if("mime_type" %in% names(res$contents)) {
-    dplyr::tbl_df(res$contents) %>% dplyr::select(path, mime_type, root, bytes, modified) # prints 25 files
-    } else {
-    dplyr::tbl_df(res$contents) %>% dplyr::select(path, is_dir, root, bytes, modified)  # prints 25 files
+  # if no limit given, make additional requests until all content retrieved
+  if (is.null(limit)) {
+    while (contents$has_more) {
+
+      # update content, append results
+      contents <- drop_list_folder_continue(contents$cursor)
+      results  <- append(results, contents$entries)
     }
   }
-  } else {
-    data.frame()
-  }
+
+  # coerce to tibble, one row per item found
+  dplyr::bind_rows(results)
+}
+
+
+#' List contents of a Dropbox folder.
+#'
+#' For internal use; drop_dir should generally be used to list files in a folder.
+#'
+#' @return a list with three elements: \code{entries}, \code{cursor}, and \code{has_more}.
+#'
+#' @references \href{https://www.dropbox.com/developers/documentation/http/documentation#files-list_folder}{API reference}
+#'
+#' @noRd
+drop_list_folder <- function(
+  path,
+  recursive = FALSE,
+  include_media_info = FALSE,
+  include_deleted = FALSE,
+  include_has_explicit_shared_members = FALSE,
+  include_mounted_folders = TRUE,
+  limit = NULL,
+  dtoken = get_dropbox_token()
+) {
+
+  if (!is.null(limit)) assertive::assert_is_numeric(limit)
+
+  url <- "https://api.dropboxapi.com/2/files/list_folder"
+
+  # this API doesn't accept "/", so don't add slashes to empty path, remove if given
+  if (path != "") path <- add_slashes(path)
+  if (path == "/") path <- ""
+
+  req <- httr::POST(
+    url = url,
+    httr::config(token = dtoken),
+    body = drop_compact(list(
+      path = path,
+      recursive = recursive,
+      include_media_info = include_media_info,
+      include_deleted = include_deleted,
+      include_has_explicit_shared_members = include_has_explicit_shared_members,
+      include_mounted_folders = include_mounted_folders,
+      limit = if (!is.null(limit)) as.integer(limit)
+    )),
+    encode = "json"
+  )
+
+  httr::stop_for_status(req)
+
+  httr::content(req)
+}
+
+
+#' Helper function for fetching additional results
+#'
+#' @noRd
+drop_list_folder_continue <- function(cursor, dtoken = get_dropbox_token()) {
```

Comment:
For `drop_list_entries` and `drop_list_folder_continue` add the @internal tag to denote internal function. Using the internal keyword removes all functions in the associated .Rd file from the documentation index and disables some of their automated tests.

Reply:
done!